File: src/main/java/uk/gov/justice/hmpps/casenotes/services/ExternalApiService.java
package uk.gov.justice.hmpps.casenotes.services;

import com.google.common.base.Joiner;
import lombok.AllArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import uk.gov.justice.hmpps.casenotes.dto.BookingIdentifier;
import uk.gov.justice.hmpps.casenotes.dto.CaseNoteFilter;
import uk.gov.justice.hmpps.casenotes.dto.CaseNoteTypeDto;
import uk.gov.justice.hmpps.casenotes.dto.NewCaseNote;
import uk.gov.justice.hmpps.casenotes.dto.NomisCaseNote;
import uk.gov.justice.hmpps.casenotes.dto.OffenderBooking;
import uk.gov.justice.hmpps.casenotes.dto.UpdateCaseNote;

import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@AllArgsConstructor
public class ExternalApiService {

    private final WebClient elite2ApiWebClient;
    private final WebClient oauthApiWebClient;
    private final WebClient elite2ClientCredentialsWebClient;

    List<CaseNoteTypeDto> getCaseNoteTypes() {
        return getCaseNoteTypes("/api/reference-domains/caseNoteTypes");
    }

    List<CaseNoteTypeDto> getUserCaseNoteTypes() {
        return getCaseNoteTypes("/api/users/me/caseNoteTypes");
    }

    private List<CaseNoteTypeDto> getCaseNoteTypes(final String url) {
        return elite2ApiWebClient.get().uri(url)
                .retrieve()
                .bodyToMono(
                        new ParameterizedTypeReference<List<CaseNoteTypeDto>>() {
                        })
                .block();
    }

    List<BookingIdentifier> getMergedIdentifiersByBookingId(final Long bookingId) {
        return elite2ClientCredentialsWebClient.get().uri("/api/bookings/{bookingId}/identifiers?type={type}", bookingId, "MERGED")
                .retrieve()
                .bodyToMono(
                        new ParameterizedTypeReference<List<BookingIdentifier>>() {
                        })
                .block();
    }

    OffenderBooking getBooking(final Long bookingId) {
        return elite2ClientCredentialsWebClient.get().uri("/api/bookings/{bookingId}?basicInfo=true", bookingId)
                .retrieve()
                .bodyToMono(OffenderBooking.class)
                .block();
    }

    String getUserFullName(final String currentUsername) {
        return oauthApiWebClient.get().uri("/api/user/{username}", currentUsername)
                .retrieve()
                .bodyToMono(
                        new ParameterizedTypeReference<Map<String, String>>() {
                        })
                .blockOptional().map(u -> u.getOrDefault("name", currentUsername)).orElse(currentUsername);
    }

    String getOffenderLocation(final String offenderIdentifier) {
        return elite2ApiWebClient.get().uri("/api/bookings/offenderNo/{offenderNo}", offenderIdentifier)
                .retrieve()
                .bodyToMono(OffenderBooking.class)
                .map(OffenderBooking::getAgencyId)
                .block();
    }

    Page<NomisCaseNote> getOffenderCaseNotes(final String offenderIdentifier, final CaseNoteFilter filter, final Pageable pageable) {

        final var paramFilter = getParamFilter(filter, pageable);
        final var url = "/api/offenders/{offenderIdentifier}/case-notes/v2?" + paramFilter;

        return elite2ApiWebClient.get().uri(url, offenderIdentifier)
                .retrieve()
                .toEntity( new ParameterizedTypeReference<RestResponsePage<NomisCaseNote>>() {})
                .map(e -> new PageImpl<NomisCaseNote>(e.getBody().getContent(), e.getBody().getPageable(), e.getBody().getTotalElements()))
                .block();
    }

    private String getParamFilter(final CaseNoteFilter filter, final Pageable pageable) {
        final var paramFilterMap = new HashMap<String, String>();

        if (StringUtils.isNotBlank(filter.getType())) {
            paramFilterMap.put("type", filter.getType());
        }

        if (StringUtils.isNotBlank(filter.getSubType())) {
            paramFilterMap.put("subType", filter.getSubType());
        }

        if (filter.getLocationId() != null) {
            paramFilterMap.put("prisonId", filter.getLocationId());
        }

        if (filter.getStartDate() != null) {
            paramFilterMap.put("from", filter.getStartDate().format(DateTimeFormatter.ISO_DATE));
        }

        if (filter.getEndDate() != null) {
            paramFilterMap.put("to", filter.getEndDate().format(DateTimeFormatter.ISO_DATE));
        }

        paramFilterMap.put("size", String.valueOf(pageable.getPageSize()));
        paramFilterMap.put("page", String.valueOf(pageable.getPageNumber()));

        final var params = paramFilterMap.isEmpty() ? null : Joiner.on("&").withKeyValueSeparator("=").join(paramFilterMap);

        final var sortParams = new StringBuilder();
        pageable.getSort().get().forEach(o -> sortParams.append("sort=").append(o.getProperty()).append(",").append(o.getDirection()));

        return params+(sortParams.length() > 0 ? "&"+sortParams : "");
    }

    NomisCaseNote createCaseNote(final String offenderIdentifier, final NewCaseNote newCaseNote) {
        return elite2ApiWebClient.post().uri("/api/offenders/{offenderNo}/case-notes", offenderIdentifier)
                .bodyValue(newCaseNote)
                .retrieve()
                .bodyToMono(NomisCaseNote.class)
                .block();
    }

    NomisCaseNote getOffenderCaseNote(final String offenderIdentifier, final long caseNoteIdentifier) {
        return elite2ApiWebClient.get().uri("/api/offenders/{offenderNo}/case-notes/{caseNoteIdentifier}", offenderIdentifier, caseNoteIdentifier)
                .retrieve()
                .bodyToMono(NomisCaseNote.class)
                .block();
    }

    NomisCaseNote amendOffenderCaseNote(final String offenderIdentifier, final long caseNoteIdentifier, final UpdateCaseNote caseNote) {
        return elite2ApiWebClient.put().uri("/api/offenders/{offenderNo}/case-notes/{caseNoteIdentifier}", offenderIdentifier, caseNoteIdentifier)
                .bodyValue(caseNote)
                .retrieve()
                .bodyToMono(NomisCaseNote.class)
                .block();
    }
}

File: src/main/java/uk/gov/justice/hmpps/casenotes/services/ExternalApiService.kt
package uk.gov.justice.hmpps.casenotes.services

import org.springframework.core.ParameterizedTypeReference
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToFlux
import org.springframework.web.reactive.function.client.bodyToMono
import reactor.core.publisher.Mono
import uk.gov.justice.hmpps.casenotes.dto.BookingIdentifier
import uk.gov.justice.hmpps.casenotes.dto.CaseNoteFilter
import uk.gov.justice.hmpps.casenotes.dto.NomisCaseNote
import uk.gov.justice.hmpps.casenotes.dto.OffenderBooking
import uk.gov.justice.hmpps.casenotes.dto.UserDetails
import uk.gov.justice.hmpps.casenotes.integrations.retryOnTransientException
import uk.gov.justice.hmpps.casenotes.notes.AmendCaseNoteRequest
import uk.gov.justice.hmpps.casenotes.notes.CreateCaseNoteRequest
import java.time.format.DateTimeFormatter

@Service
class ExternalApiService(
  private val elite2ApiWebClient: WebClient,
  private val oauthApiWebClient: WebClient,
  private val elite2ClientCredentialsWebClient: WebClient,
) {

  fun getMergedIdentifiersByBookingId(bookingId: Long): List<BookingIdentifier> =
    elite2ClientCredentialsWebClient.get()
      .uri("/api/bookings/{bookingId}/identifiers?type={type}", bookingId, "MERGED")
      .retrieve()
      .bodyToFlux<BookingIdentifier>()
      .collectList()
      .block()!!

  fun getBooking(bookingId: Long): OffenderBooking =
    elite2ClientCredentialsWebClient.get().uri("/api/bookings/{bookingId}?basicInfo=true", bookingId)
      .retrieve()
      .bodyToMono<OffenderBooking>()
      .retryOnTransientException()
      .block()!!

  fun getUserDetails(currentUsername: String): UserDetails? =
    oauthApiWebClient.get().uri("/api/user/{username}", currentUsername)
      .exchangeToMono {
        if (it.statusCode() == HttpStatus.NOT_FOUND) {
          Mono.empty()
        } else {
          it.bodyToMono<UserDetails>()
        }
      }
      .retryOnTransientException()
      .block()

  fun getOffenderCaseNotes(
    offenderIdentifier: String,
    filter: CaseNoteFilter,
    pageable: Pageable,
  ): Page<NomisCaseNote> {
    val paramFilter = getParamFilter(filter, pageable)
    val url = "/api/offenders/{offenderIdentifier}/case-notes/v2?$paramFilter"
    return elite2ApiWebClient.get()
      .uri(url, offenderIdentifier, *filter.getTypesAndSubTypes().asPrisonApiParams().toTypedArray())
      .retrieve()
      .toEntity(object : ParameterizedTypeReference<RestResponsePage<NomisCaseNote>>() {})
      .retryOnTransientException()
      .map { e: ResponseEntity<RestResponsePage<NomisCaseNote>> ->
        PageImpl(
          e.body!!.content,
          e.body!!.pageable,
          e.body!!.totalElements,
        )
      }
      .block()!!
  }

  private fun getParamFilter(filter: CaseNoteFilter, pageable: Pageable): String {
    val paramFilterMap = mutableMapOf(
      "page" to pageable.pageNumber.toString(),
      "size" to pageable.pageSize.toString(),
    ).apply {
      filter.locationId?.let { this["prisonId"] = filter.locationId }
      filter.startDate?.let { this["from"] = filter.startDate.format(DateTimeFormatter.ISO_DATE) }
      filter.endDate?.let { this["to"] = filter.endDate.format(DateTimeFormatter.ISO_DATE) }
    }
    val params = paramFilterMap.entries.joinToString(separator = "&") { (key, value) -> "$key=$value" }
    val sortParams = pageable.sort.map {
      val mappedProperty = if (it.property == "creationDateTime") "createDatetime" else it.property
      "sort=$mappedProperty,${it.direction}"
    }
      .joinToString(separator = "&", prefix = "&")

    if (filter.getTypesAndSubTypes().isNotEmpty()) {
      val typeSubTypesParams = filter.getTypesAndSubTypes().asPrisonApiParams()
        .joinToString(separator = "&", prefix = "&") { "typeSubTypes={typeSubTypes}" }
      return "$params$typeSubTypesParams$sortParams"
    }

    return "$params$sortParams"
  }

  fun createCaseNote(offenderIdentifier: String, newCaseNote: CreateCaseNoteRequest): NomisCaseNote =
    elite2ApiWebClient.post().uri("/api/offenders/{offenderNo}/case-notes", offenderIdentifier)
      .bodyValue(newCaseNote)
      .retrieve()
      .bodyToMono<NomisCaseNote>()
      .retryOnTransientException()
      .block()!!

  fun getOffenderCaseNote(offenderIdentifier: String, caseNoteIdentifier: Long): NomisCaseNote =
    elite2ApiWebClient.get()
      .uri("/api/offenders/{offenderNo}/case-notes/{caseNoteIdentifier}", offenderIdentifier, caseNoteIdentifier)
      .retrieve()
      .bodyToMono<NomisCaseNote>()
      .retryOnTransientException()
      .block()!!

  fun amendOffenderCaseNote(
    offenderIdentifier: String,
    caseNoteIdentifier: Long,
    caseNote: AmendCaseNoteRequest,
  ): NomisCaseNote = elite2ApiWebClient.put()
    .uri("/api/offenders/{offenderNo}/case-notes/{caseNoteIdentifier}", offenderIdentifier, caseNoteIdentifier)
    .bodyValue(caseNote)
    .retrieve()
    .bodyToMono<NomisCaseNote>()
    .retryOnTransientException()
    .block()!!

  private fun Map<String, Set<String>>.asPrisonApiParams(): List<String> =
    entries.flatMap { e -> if (e.value.isEmpty()) setOf(e.key) else e.value.map { "${e.key}+$it" } }
}
